src/styles/style-helpers.ts

```diff
@@ -25,6 +26,133 @@ function isTheme(args: ThemeArgs): args is Theme {
 export const getTheme = (args: ThemeArgs): Theme =>
   isTheme(args) ? args : args.theme;
 
+type Spacing = keyof Theme['spacings'] | null;
+
+const spacingString = (size: Spacing, theme: Theme) => {
+  if (!size) {
+    if (process.env.NODE_ENV !== 'production') {
+      console.warn(
+        'A single `null` value was passed to the spacing mixin. This has no effect since `null` represents no set margin, not zero margin. If you are trying to reset the spacing, use custom styles instead.',
+      );
+    }
+    return null;
+  }
+  return css({ margin: theme.spacings[size] });
+};
+
+const spacingObject = (
+  size: { top?: Spacing; bottom?: Spacing; right?: Spacing; left?: Spacing },
+  theme: Theme,
+) => {
+  const margins: {
+    marginTop?: string;
+    marginBottom?: string;
+    marginRight?: string;
+    marginLeft?: string;
+  } = {};
+  if (size.top) {
+    margins.marginTop = theme.spacings[size.top];
+  }
+
+  if (size.bottom) {
+    margins.marginBottom = theme.spacings[size.bottom];
+  }
+
+  if (size.right) {
+    margins.marginRight = theme.spacings[size.right];
+  }
+
+  if (size.left) {
+    margins.marginLeft = theme.spacings[size.left];
+  }
+
+  return css(margins);
+};
+const spacingArray = (
+  size: [Spacing, Spacing?, Spacing?, Spacing?],
+  theme: Theme,
+) => {
+  if (size.length === 1) {
+    if (!size[0]) {
+      if (process.env.NODE_ENV !== 'production') {
+        console.warn(
+          'A single `null` value was passed to the spacing mixin. This has no effect since `null` represents no set margin, not zero margin. If you are trying to reset the spacing, use custom styles instead.',
+        );
+      }
+      return null;
+    }
+    return css({ margin: theme.spacings[size[0]] });
+  }
+  if (size.length === 2) {
+    const margins: {
+      marginTop?: string;
+      marginBottom?: string;
+      marginRight?: string;
+      marginLeft?: string;
+    } = {};
+    if (size[0]) {
+      margins.marginTop = theme.spacings[size[0]];
+      margins.marginBottom = theme.spacings[size[0]];
+    }
+
+    if (size[1]) {
```

Comment:
Wouldn't both of these conditions always be true as long as `size.length === 2`?

Reply:
Oh I guess you could pass `[null, Spacing]`

Reply:
I'm not sure if I'm a fan of the pattern. It seems like it could produce some unexpected behaviour.

For example, maybe someone is passing `[null, Spacing]` with the intention of no spacing on the top/bottom, and `Spacing` on the left/right.

Reply:
I included this pattern in the [RFC](https://github.com/sumup-oss/circuit-ui/issues/534) as a shorthand for vertical/horizontal spacing. It's more ergonomic than `spacing({ top: 'size', bottom: 'size' })`.

I modeled the behavior for the array syntax after the CSS shorthands. The only difference is that instead of setting a _0_ margin, the spacing mixin sets _no_ margin when passed `null` to avoid overriding existing margins (useful when styling Circuit UI's grid components). Is this the difference you're referring to? 

An alternative would be to support `spacing({ horizontal: 'size' })`. What happens if you use `spacing({ horizontal: 'size', right: 'size' })` — which takes precedence? 

> And if we do handle it, then why wouldn't we handle the same potential null values when length === 3 for example?

We do. The check happens in the `spacingObject` function. We could rewrite the `size.length === 3` case to also use `spacingObject` for consistency:

```ts
if (size.length === 2) {
    return spacingObject(
      {
        top: size[0],
        right: size[1],
        bottom: size[0],
        left: size[1],
      },
      theme,
    );
  }
```

Reply:
> I personally find spacing(['size', null]) more ergonomic than spacing({ top: 'size', bottom: 'size' })

I know the syntax is there in the RFC I'd just like to point out that in most cases, clarity > ergonomics. I'm personally not even a fan of shorthands, so I'll admit a bit of bias. My brain sees `spacing([ null, Spacing ])` as "shorthand for `0` on the left and right, `Spacing` on the top and bottom". You have to have some kind of understanding with the way that JavaScript/CSS play together to know that `null` is not a value but a "non value".

Reply:
Thanks for the feedback! 

> in most cases, clarity > ergonomics

That's a very good point. A couple of questions for you (@mcntsh) and @robinmetral:

1. Would you get rid of the array syntax entirely then?
2. Would you get rid of the `null` (non) value support?
3. Would you add support for actual `0` pixel spacing, ie. a spacing reset?

Reply:
Thanks for the details!

My personal preference would be:

1. Yes
2. Doesn't apply if we don't have the array syntax
3. I think this would be a good idea, to eventually replace things like the noMargin prop - we could introduce a new theme.spacings value of "none" for example, but this is a different discussion

About 1., like I mentioned in https://github.com/sumup-oss/circuit-ui/pull/832#discussion_r587400301 I would personally only expose a single API for granular spacing controls, because it would ensure a clearer and more consistent usage or the mixin.

On the other hand I agree that `spacing({ top: 'size', bottom: 'size' })` to get vertical spacing is a little bit verbose, and since the array pattern was already presented in an RFC, we can also consider it approved and move forward with this (great) implementation.

So, as you wish 🙂

Reply:
Thanks for your input! I'd be in favor of removing the array syntax, for the time being. We can always add it back later.

> I think this would be a good idea, to eventually replace things like the `noMargin` prop - we could introduce a new `theme.spacings` value of "none" for example, but this is a different discussion

The default outer spacing will be removed in the next major release of Circuit UI (see https://github.com/sumup-oss/circuit-ui/issues/534). As part of the migration strategy, the `noMargin` prop will be enforced in the next minor release and removed entirely in the next major release.